pkg/gardenlet/operation/shoot/shoot.go

```diff
@@ -545,6 +546,36 @@ func sortByIPFamilies(ipfamilies []gardencorev1beta1.IPFamily, cidr []net.IPNet)
 	return result
 }
 
+func getPrimaryCIDRs(cidrs []net.IPNet, ipFamilies []gardencorev1beta1.IPFamily) []net.IPNet {
+	var result []net.IPNet
+	isIPv4 := ipFamilies[0] == gardencorev1beta1.IPFamilyIPv4
+	for _, c := range cidrs {
+		if (isIPv4 && c.IP.To4() != nil) || (!isIPv4 && c.IP.To4() == nil) {
+			result = append(result, c)
+		}
+	}
+	return result
+}
+
+// CheckDualStackMigrateNetworks checks if the shoot should be migrated to dual-stack networking and sets the shoot status accoridingly.
+func (s *Shoot) CheckDualStackMigrateNetworks(ctx context.Context, gardenClient client.Client, clock clock.Clock) error {
+	shoot := s.GetInfo()
+
+	condition := v1beta1helper.GetCondition(shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady)
+	if condition == nil && len(shoot.Spec.Networking.IPFamilies) == 2 && shoot.Status.Networking != nil && len(shoot.Status.Networking.Nodes) == 1 {
+		err := s.UpdateInfoStatus(ctx, gardenClient, true, func(shoot *gardencorev1beta1.Shoot) error {
+			condition := v1beta1helper.GetOrInitConditionWithClock(clock, shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady)
+			condition = v1beta1helper.UpdatedConditionWithClock(clock, condition, gardencorev1beta1.ConditionFalse, "DualStackMigration", "The shoot is migrating to dual-stack networking.")
+			shoot.Status.Constraints = v1beta1helper.MergeConditions(shoot.Status.Constraints, condition)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
Would this function not fit better into the `Botanist`? This would reduce the amount of parameters as well.